pkg/causality/conflict_detector.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package causality
+
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/pingcap/tiflow/engine/pkg/containers"
+	"github.com/pingcap/tiflow/pkg/causality/internal"
+)
+
+// ConflictDetector implements a logic that dispatches transaction
+// to different workers in a way that transactions modifying the same
+// keys are never executed concurrently and have their original orders
+// preserved.
+type ConflictDetector[Worker worker[Txn], Txn txnEvent] struct {
+	workers []Worker
+	slots   *internal.Slots[*internal.Node]
```

Comment:
could you add some comments about the slots,finishedTxnQuene and resolvedTxnQueue.